README.md
[![Releases](https://img.shields.io/github/release/Telegram-FOSS-Team/Telegram-FOSS.svg)](https://github.com/Telegram-FOSS-Team/Telegram-FOSS/releases/latest)
[![Discussions](https://img.shields.io/badge/Offtopics-Telegram-blue.svg)](https://t.me/joinchat/AAAAAEFaw9LIC-VgKVCevg)

[![Donate](https://liberapay.com/assets/widgets/donate.svg)](https://liberapay.com/Telegram-FOSS/) or <img src="https://en.bitcoin.it/w/images/en/c/cb/BC_Logotype.png" alt="Bitcoin" height="25px" /> `1McafEgMvqAVujNLtcJumZHxp2UfaNByqs`

# Telegram-FOSS

[Telegram](https://telegram.org) is a messaging app with a focus on speed and security. It’s superfast, simple and free.

This is an unofficial, FOSS-friendly fork of the original [Telegram App for Android](https://github.com/DrKLO/Telegram).

This version of Telegram is available on FDroid:

[<img src="https://f-droid.org/badge/get-it-on.png"
      alt="Get it on F-Droid"
      height="80">](https://f-droid.org/app/org.telegram.messenger)

## Current Maintainers

- [thermatk](https://github.com/thermatk)
- [Bubu](https://github.com/Bubu)
- you? :)

## Contributors

- [slp](https://github.com/slp)
- [Sudokamikaze](https://github.com/Sudokamikaze)
- [l2dy](https://github.com/l2dy)
- [maximgrafin](https://github.com/maximgrafin)

## Discussion

Join the [Telegram-FOSS Offtopics group](https://t.me/joinchat/AAAAAEFaw9LIC-VgKVCevg)

You can also join `#telegram-foss:matrix.org` via [matrix](https://matrix.to/#/#telegram-foss:matrix.org) or `#telegram-foss` on freenode.

## Changes:

*Replacement of non-FOSS, untrustworthy or suspicious binaries or source code:*
- Do location sharing with OpenStreetMap(osmdroid) instead of Google Maps
- Use Twemoji emoji set instead of Apple's emoji
- Google Play Services GCM replaced with always-on Telegram's push service
 - Has to show a notification on Oreo+, ask Google
- BoringSSL prebuilts are replaced with BoringSSL built at compile time
- **SECURITY:** Old FFmpeg prebuilts are replaced with the newest upstream source code built at compile time
- **SECURITY:** Bundled libWebP is updated

*Removal of non-FOSS, untrustworthy or suspicious binaries or source code and their functionality:*
- Google Vision face detection
- Google Wallet and Android Pay integration
- HockeyApp crash reporting and self-updates

*Other:*
- Allow to set a proxy before login
- Added the ability to parse locations from intents containing a `geo:<lat>,<lon>,<zoom>` string
- Force static map previews from Telegram

## Versioning

This repository contains tags to make tracking versions easier.

Versions are in form "v$UPSTREAM$RELEASE" where:

* $UPSTREAM is the public, visible version of upstream.
* $RELEASE is a letter ([a-z]) indicating minor releases between official versions (sometimes, upstream is updated without relating the changes to an specific version).

## API, Protocol documentation

Telegram API manuals: https://core.telegram.org/api

MTproto protocol manuals: https://core.telegram.org/mtproto

## Building

**Important:**
1. You need the [Android NDK, both revision 16b and 14b](https://developer.android.com/ndk/downloads/older_releases), and [Ninja](https://ninja-build.org/) to build the apk.

2. Don't forget to include the submodules when you clone:
      - `git clone --recursive https://github.com/Telegram-FOSS-Team/Telegram-FOSS.git`

3. Build native FFmpeg and BoringSSL dependencies:
      - Go to the `TMessagesProj/jni` folder and execute the following (define the paths to your NDKs and Ninja):

      ```
      export NDK=[PATH_TO_NDK_R14B]
      ./build_ffmpeg_android.sh
      export NDK=[PATH_TO_NDK_R16B]
      export NINJA_PATH=[PATH_TO_NINJA]
      ./build_boringssl.sh
      ```

4. For the whole project set NDK to r16b

5. If you want to publish a modified version of Telegram:
      - You should get **your own API key** here: https://core.telegram.org/api/obtaining_api_id and create a file called `API_KEYS` in the source root directory.
        The contents should look like this:
        ```
        APP_ID = 12345
        APP_HASH = aaaaaaaabbbbbbccccccfffffff001122
        ```
      - Do not use the name Telegram and the standard logo (white paper plane in a blue circle) for your app — or make sure your users understand that it is unofficial
      - Take good care of your users' data and privacy
      - **Please remember to publish your code too in order to comply with the licenses**

6. Building under Windows may be problematic and won't be fixed

The project can be built with Android Studio or from the command line with gradle:

`./gradlew assembleFatRelease`

# DIGITAL RESISTANCE

![DIGITALRESISTANCE](/DigitalResistance.jpg?raw=true "DIGITALRESISTANCE")



Changelog.md
### v4.9
- Telegram Passport – a unified authorization method for services that require personal identification.
- Store your identity documents and personal data in the Telegram cloud using End-to-End Encryption.
- Instantly share your data with services that require real-world ID (finance, ICOs, etc.).
- For a real-life implementation, see www.ePayments.com/tg – the first electronic payments system to integrate Telegram Passport.
- **Telegram FOSS:**
  - FFmpeg 4.0.2
  - libwebp 1.0.0
  - now using BoringSSL
  - suggested locations and search are back
  - osmdroid 6.0.2 and https for map tiles
  - various cleanups and fixes

### v4.6c
- **Telegram FOSS:**
  - Don't send photo caption twice #247
  - Zoom set to 4 by default for location sharing
  - Update FFmpeg to 4.0.1

#### v4.8.9 (not released as FOSS version)
- Updated the registration process. If you live in the UK or EU, you must be 16 years or older to use Telegram.
- When signing up for Telegram, you accept our Privacy Policy (telegram.org/privacy).
- You can now stop updating your contacts and delete your synced contacts in Privacy & Security settings.
- If you enabled link previews in Secret Chats, you can now disable them in Privacy & Security settings.

### v4.6b
- **Telegram FOSS:**
  - Allow to set a proxy before login #230
  - Use MS DF 2 PWN RKN. Getting connected to Telegram behind a country-wide block may require you to stare on your login screen for some seconds.
  - Update FFmpeg to 3.4.2
  - Update OpenSSL to 1.0.2o
  - Update SQLite to 3.23.1

#### v4.8.5 (not released as FOSS version)
- Discover new stickers. Type one emoji to see suggestions from popular sticker sets. Suggestions from the sticker sets you've added will come first.
- Search for Stickers. Scroll up in the sticker panel and use the new search field to quickly locate your sticker sets or discover new ones.
- Multi-shot sending. Take and send multiple photos and videos one after another.

#### v4.8.2 (not released as FOSS version)
- Fixed sunrise and sunset time issues for Auto-Night Mode.
- Fixed badge counter for Nova Launcher.
- Fixed round videos recording on some devices.

#### v4.8 (not released as FOSS version)
- Streaming for videos. Start watching any newly uploaded video instantly without having to fully download it first.
- Auto-Night Mode. Automatically switch to the dark version of the interface after nightfall or in low-light conditions.
- Telegram Login widget. Log in to other websites and services using your Telegram account.

#### v4.7 (not released as FOSS version)
- Quickly switch between different Telegram accounts if you use multiple phone numbers.
- Swipe left on any message to reply to it.

### v4.6a
- **Telegram FOSS:**
  - version bump to rebuild with NDK r14b
  - fixes unsupported content for url sharing #200

### v4.6
- New granular settings for auto-downloading media.
- Link previews for Instagram posts and tweets with multiple photos will now show all the media as an album.
- Embeddable HTML-widget for messages in public channels and groups (available when viewing t.me links to messages in web-browsers).
- Added support for albums to Secret Chats.
- Added full support for MTProto 2.0.
- **Telegram FOSS:**
  - emojiOne replaced with Twemoji

### v4.2.1b
- **Telegram FOSS:**
  - update OpenSSL to 1.0.2n
  - fix videos without sound
  - fix image sharing bug (sorry for that one!)

### v4.2.1a
- **Telegram FOSS:**
  - **SECURITY:** update OpenSSL to 1.0.2m
  - **SECURITY:** update FFmpeg to 3.4
  - **SECURITY:** update SQLite and libWebP to latest versions.
  - Build native code for armv6 devices
  - Dropped the unused Google Breakpad library
  - Can and should now be built with NDK r15c
  - Added location sharing through "geo:" intents

#### v4.5 (not released as FOSS version)
- Grouped Photos. Group media into albums when sharing multiple photos and videos. Choose the exact order of media you send.
- Saved Messages. Bookmark messages by forwarding them to “Saved Messages”.
- Better Search. Find bots and public channels faster by typing their titles in Search. Popular bots and channels are shown first.
- Pinned Messages. Pin messages in your channels to make important announcements more visible.

#### v4.4 (not released as FOSS version)
- Share your location with friends in real time with the new Live Locations.
- Control whether new members in supergroups can see earlier message history.
- Easily recognize messages from group admins by the new ‘admin’ badge.
- Listen to audio files with more comfort using the redesigned player.
- Added French, Malay, Indonesian, Russian, and Ukrainian languages.
- Suggest alternative translations using our new localization platform – translations.telegram.org

#### v4.3 (not released as FOSS version)
- Groups with unread mentions and replies are now marked with an '@' badge in the chats list.
- Navigate new mentions and replies in a group using the new '@' button.
- Mark stickers as Favorite to quickly access them from the redesigned sticker panel.
- Invite friends to Telegram using the new streamlined interface.
- Add an official sticker set for your group to be used by all members without adding (100+ member groups only).
- Numerous design and UI improvements.

#### v4.2.2 (not released as FOSS version)
- Add emoji to a message by typing ‘:’ + keyword. :relieved : :satisfied : :smirk :
- Search through messages of a particular user in any group. To do this, tap '...' in the top right corner when in a group > Search > tap the new 'Search by member' icon in the bottom right corner.
- While searching, select a user to browse all of her messages in the group or add a keyword to narrow down search results.

### v4.2.1
- Send self-destructing photos and videos to any one-on-one chats (use the clock icon in the media picker to set a timer).
- Edit photos even quicker with the improved photo editor.
- Add a bio to your profile (in Settings) so that people in large group chats know who you are.

TMessagesProj/jni/build_ffmpeg_clang.sh
#!/bin/bash
set -e
function build_one {
    CC="${CROSS_PREFIX}clang"
    CXX="${CROSS_PREFIX}clang++"
    AS="${CROSS_PREFIX}clang"
    AR="${CROSS_PREFIX}ar"
    LD="${CROSS_PREFIX}ld"
    NM="${CROSS_PREFIX}nm"
    STRIP="${CROSS_PREFIX}strip"

	echo "Cleaning..."
	rm -f config.h
	make clean
	rm -rf ${TOOLCHAIN_PREFIX}
	
	echo "Toolchain..."
	python $NDK/build/tools/make_standalone_toolchain.py \
	--arch ${ARCH} \
	--api ${ANDROID_API} \
	--stl libc++ \
	--install-dir=${TOOLCHAIN_PREFIX}

	echo "Configuring..."

	./configure \
	--nm=${NM} \
	--ar=${AR} \
	--as=${CROSS_PREFIX}gcc \
	--strip=${STRIP} \
	--cc=${CC} \
	--cxx=${CXX} \
	--enable-stripping \
	--arch=$ARCH \
	--target-os=linux \
	--enable-cross-compile \
	--x86asmexe=$NDK/prebuilt/$BUILD_PLATFORM/bin/yasm \
	--prefix=$PREFIX \
	--enable-pic \
	--disable-shared \
	--enable-static \
    --enable-asm \
    --enable-inline-asm \
    --enable-x86asm \
	--cross-prefix=$CROSS_PREFIX \
	--sysroot=$SYSROOT \
	--extra-cflags="-Wl,-Bsymbolic -Os -DCONFIG_LINUX_PERF=0 -DANDROID $OPTIMIZE_CFLAGS -fPIE -pie --static -fPIC" \
	--extra-ldflags="-Wl,-Bsymbolic -Wl,-rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -nostdlib -lc -lm -ldl -fPIC" \
	\
	--enable-version3 \
	--enable-gpl \
	\
	--disable-linux-perf \
	\
	--disable-doc \
	--disable-htmlpages \
	--disable-avx \
	\
	--disable-everything \
	--disable-network \
	--disable-zlib \
	--disable-avfilter \
	--disable-avdevice \
	--disable-postproc \
	--disable-debug \
	--disable-programs \
	--disable-network \
	--disable-ffplay \
	--disable-ffprobe \
	--disable-swscale \
	--disable-postproc \
	--disable-avdevice \
	\
	--enable-runtime-cpudetect \
	--enable-pthreads \
	--enable-avresample \
	--enable-protocol=file \
	--enable-decoder=h264 \
	--enable-decoder=mpeg4 \
	--enable-decoder=gif \
	--enable-decoder=alac \
	--enable-demuxer=mov \
	--enable-demuxer=gif \
	--enable-hwaccels \
	$ADDITIONAL_CONFIGURE_FLAG

	#echo "continue?"
	#read
	make -j$COMPILATION_PROC_COUNT
	make install
}

function setCurrentPlatform {

    PLATFORM="$(uname -s)"
    case "${PLATFORM}" in
        Darwin*)
                    BUILD_PLATFORM=darwin-x86_64
                    COMPILATION_PROC_COUNT=`sysctl -n hw.physicalcpu`
                    ;;
        Linux*)
                    BUILD_PLATFORM=linux-x86_64
                    COMPILATION_PROC_COUNT=$(nproc)
                    ;;
        *)
                    echo -e "\033[33mWarning! Unknown platform ${PLATFORM}! falling back to linux-x86_64\033[0m"
                    BUILD_PLATFORM=linux-x86_64
                    COMPILATION_PROC_COUNT=1
                    ;;
    esac

    echo "build platform: ${BUILD_PLATFORM}"
    echo "parallel jobs: ${COMPILATION_PROC_COUNT}"

}

function checkPreRequisites {

    if ! [ -d "ffmpeg" ] || ! [ "$(ls -A ffmpeg)" ]; then
        echo -e "\033[31mFailed! Submodule 'ffmpeg' not found!\033[0m"
        echo -e "\033[31mTry to run: 'git submodule init && git submodule update'\033[0m"
        exit
    fi

    if [ -z "$NDK" -a "$NDK" == "" ]; then
        echo -e "\033[31mFailed! NDK is empty. Run 'export NDK=[PATH_TO_NDK]'\033[0m"
        exit
    fi
}

setCurrentPlatform
checkPreRequisites

# TODO: fix env variable for NDK
# NDK=/opt/android-sdk/ndk-bundle

cd ffmpeg

BASEDIR=`pwd`
TOOLCHAIN_PREFIX=${BASEDIR}/toolchain-android

## common
SYSROOT=${TOOLCHAIN_PREFIX}/sysroot

#x86_64
ANDROID_API=21
PREBUILT=$NDK/toolchains/x86_64-4.9/prebuilt/$BUILD_PLATFORM
PLATFORM=$NDK/platforms/android-21/arch-x86_64
CROSS_PREFIX=${TOOLCHAIN_PREFIX}/bin/x86_64-linux-android-
ARCH=x86_64
CPU=x86_64
PREFIX=./android/$CPU
ADDITIONAL_CONFIGURE_FLAG="--disable-asm"
build_one

#arm64-v8a
ANDROID_API=21
PREBUILT=$NDK/toolchains/aarch64-linux-android-4.9/prebuilt/$BUILD_PLATFORM
PLATFORM=$NDK/platforms/android-21/arch-arm64
CROSS_PREFIX=${TOOLCHAIN_PREFIX}/bin/aarch64-linux-android-
ARCH=arm64
CPU=arm64-v8a
OPTIMIZE_CFLAGS=
PREFIX=./android/$CPU
ADDITIONAL_CONFIGURE_FLAG="--enable-neon --enable-optimizations"
#build_one


#arm v7n
ANDROID_API=16
PREBUILT=$NDK/toolchains/arm-linux-androideabi-4.9/prebuilt/$BUILD_PLATFORM
PLATFORM=$NDK/platforms/android-16/arch-arm
CROSS_PREFIX=${TOOLCHAIN_PREFIX}/bin/arm-linux-androideabi-
ARCH=arm
CPU=armv7-a
OPTIMIZE_CFLAGS="-marm -march=$CPU"
PREFIX=./android/$CPU
ADDITIONAL_CONFIGURE_FLAG="--enable-neon"
#build_one

#x86 platform
ANDROID_API=16
PREBUILT=$NDK/toolchains/x86-4.9/prebuilt/$BUILD_PLATFORM
PLATFORM=$NDK/platforms/android-16/arch-x86
CROSS_PREFIX=${TOOLCHAIN_PREFIX}/bin/i686-linux-android-
ARCH=x86
CPU=i686
OPTIMIZE_CFLAGS="-march=$CPU"
PREFIX=./android/$CPU
ADDITIONAL_CONFIGURE_FLAG="--disable-x86asm --disable-inline-asm --disable-asm"
#build_one


TMessagesProj/jni/build_ffmpeg_android.sh
#!/bin/bash

function build_one {

echo "Cleaning..."
rm config.h
make clean

echo "Configuring..."

./configure \
--cc=$CC \
--nm=$NM \
--enable-stripping \
--arch=$ARCH \
--target-os=linux \
--enable-cross-compile \
--x86asmexe=$NDK/prebuilt/$BUILD_PLATFORM/bin/yasm \
--prefix=$PREFIX \
--enable-pic \
--disable-shared \
--enable-static \
--enable-asm \
--enable-inline-asm \
--cross-prefix=$CROSS_PREFIX \
--sysroot=$PLATFORM \
--extra-cflags="-Wl,-Bsymbolic -Os -DCONFIG_LINUX_PERF=0 -DANDROID $OPTIMIZE_CFLAGS -fPIE -pie --static -fPIC" \
--extra-ldflags="-Wl,-Bsymbolic -Wl,-rpath-link=$PLATFORM/usr/lib -L$PLATFORM/usr/lib -nostdlib -lc -lm -ldl -fPIC" \
--extra-libs="-lgcc" \
\
--enable-version3 \
--enable-gpl \
\
--disable-linux-perf \
\
--disable-doc \
--disable-htmlpages \
--disable-avx \
\
--disable-everything \
--disable-network \
--disable-zlib \
--disable-avfilter \
--disable-avdevice \
--disable-postproc \
--disable-debug \
--disable-programs \
--disable-network \
--disable-ffplay \
--disable-ffprobe \
--disable-swscale \
--disable-postproc \
--disable-avdevice \
\
--enable-runtime-cpudetect \
--enable-pthreads \
--enable-avresample \
--enable-protocol=file \
--enable-decoder=h264 \
--enable-decoder=mpeg4 \
--enable-decoder=gif \
--enable-decoder=alac \
--enable-demuxer=mov \
--enable-demuxer=gif \
--enable-hwaccels \
--enable-runtime-cpudetect \
$ADDITIONAL_CONFIGURE_FLAG

#echo "continue?"
#read
make -j$COMPILATION_PROC_COUNT
make install

}

function setCurrentPlatform {

    PLATFORM="$(uname -s)"
    case "${PLATFORM}" in
        Darwin*)
                    BUILD_PLATFORM=darwin-x86_64
                    COMPILATION_PROC_COUNT=`sysctl -n hw.physicalcpu`
                    ;;
        Linux*)
                    BUILD_PLATFORM=linux-x86_64
                    COMPILATION_PROC_COUNT=$(nproc)
                    ;;
        *)
                    echo -e "\033[33mWarning! Unknown platform ${PLATFORM}! falling back to linux-x86_64\033[0m"
                    BUILD_PLATFORM=linux-x86_64
                    COMPILATION_PROC_COUNT=1
                    ;;
    esac

    echo "build platform: ${BUILD_PLATFORM}"
    echo "parallel jobs: ${COMPILATION_PROC_COUNT}"

}

function checkPreRequisites {

    if ! [ -d "ffmpeg" ] || ! [ "$(ls -A ffmpeg)" ]; then
        echo -e "\033[31mFailed! Submodule 'ffmpeg' not found!\033[0m"
        echo -e "\033[31mTry to run: 'git submodule init && git submodule update'\033[0m"
        exit
    fi

    if [ -z "$NDK" -a "$NDK" == "" ]; then
        echo -e "\033[31mFailed! NDK is empty. Run 'export NDK=[PATH_TO_NDK]'\033[0m"
        exit
    fi
}

setCurrentPlatform
checkPreRequisites

# TODO: fix env variable for NDK
# NDK=/opt/android-sdk/ndk-bundle

cd ffmpeg

#x86_64
PREBUILT=$NDK/toolchains/x86_64-4.9/prebuilt/$BUILD_PLATFORM
PLATFORM=$NDK/platforms/android-21/arch-x86_64
LD=$PREBUILT/bin/x86_64-linux-android-ld
AR=$PREBUILT/bin/x86_64-linux-android-ar
NM=$PREBUILT/bin/x86_64-linux-android-nm
GCCLIB=$PREBUILT/lib/gcc/x86_64-linux-android/4.9.x/libgcc.a
CC=$PREBUILT/bin/x86_64-linux-android-gcc
CROSS_PREFIX=$PREBUILT/bin/x86_64-linux-android-
ARCH=x86_64
CPU=x86_64
PREFIX=./android/$CPU
ADDITIONAL_CONFIGURE_FLAG="--disable-mmx --disable-inline-asm"
#build_one

#arm64-v8a
PREBUILT=$NDK/toolchains/aarch64-linux-android-4.9/prebuilt/$BUILD_PLATFORM
PLATFORM=$NDK/platforms/android-21/arch-arm64
LD=$PREBUILT/bin/aarch64-linux-android-ld
AR=$PREBUILT/bin/aarch64-linux-android-ar
NM=$PREBUILT/bin/aarch64-linux-android-nm
GCCLIB=$PREBUILT/lib/gcc/aarch64-linux-android/4.9.x/libgcc.a
CC=$PREBUILT/bin/aarch64-linux-android-gcc
CROSS_PREFIX=$PREBUILT/bin/aarch64-linux-android-
ARCH=arm64
CPU=arm64-v8a
OPTIMIZE_CFLAGS=
PREFIX=./android/$CPU
ADDITIONAL_CONFIGURE_FLAG="--enable-neon --enable-optimizations"
build_one


#arm v7n
PREBUILT=$NDK/toolchains/arm-linux-androideabi-4.9/prebuilt/$BUILD_PLATFORM
PLATFORM=$NDK/platforms/android-16/arch-arm
LD=$PREBUILT/bin/arm-linux-androideabi-ld
AR=$PREBUILT/bin/arm-linux-androideabi-ar
NM=$PREBUILT/bin/arm-linux-androideabi-nm
GCCLIB=$PREBUILT/lib/gcc/arm-linux-androideabi/4.9.x/libgcc.a
CC=$PREBUILT/bin/arm-linux-androideabi-gcc
CROSS_PREFIX=$PREBUILT/bin/arm-linux-androideabi-
ARCH=arm
CPU=armv7-a
OPTIMIZE_CFLAGS="-marm -march=$CPU"
PREFIX=./android/$CPU
ADDITIONAL_CONFIGURE_FLAG=--enable-neon
build_one

#x86 platform
PREBUILT=$NDK/toolchains/x86-4.9/prebuilt/$BUILD_PLATFORM
PLATFORM=$NDK/platforms/android-16/arch-x86
LD=$PREBUILT/bin/i686-linux-android-ld
AR=$PREBUILT/bin/i686-linux-android-ar
NM=$PREBUILT/bin/i686-linux-android-nm
GCCLIB=$PREBUILT/lib/gcc/i686-linux-android/4.9.x/libgcc.a
CC=$PREBUILT/bin/i686-linux-android-gcc
CROSS_PREFIX=$PREBUILT/bin/i686-linux-android-
ARCH=x86
CPU=i686
OPTIMIZE_CFLAGS="-march=$CPU"
PREFIX=./android/$CPU
ADDITIONAL_CONFIGURE_FLAG="--disable-mmx --disable-yasm"
build_one


TMessagesProj/src/main/java/org/telegram/messenger/NotificationsService.java
/*
 * This is the source code of Telegram for Android v. 1.3.x.
 * It is licensed under GNU GPL v. 2 or later.
 * You should have received a copy of the license in this archive (see LICENSE).
 *
 * Copyright Nikolai Kudashov, 2013-2017.
 */

package org.telegram.messenger;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;
import android.support.v4.app.NotificationCompat;

public class NotificationsService extends Service {

    @Override
    public void onCreate() {
        ApplicationLoader.postInitApplication();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            String CHANNEL_ID = "push_service_channel";
            NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
            NotificationChannel channel = new NotificationChannel(CHANNEL_ID,"Push Notifications Service",NotificationManager.IMPORTANCE_DEFAULT);
            notificationManager.createNotificationChannel(channel);
            Notification notification = new NotificationCompat.Builder(this, CHANNEL_ID)
                    .setSmallIcon(R.drawable.notification)
                    .setContentTitle("Push Notifications")
                    .setContentText("Google wants us to use FCM. No way!").build();
            startForeground(9999,notification);
        }
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    public void onDestroy() {
        // Telegram-FOSS: unconditionally enable push service
        Intent intent = new Intent("org.telegram.start");
        sendBroadcast(intent);
    }
}


TMessagesProj/jni/build_boringssl.sh
#!/bin/bash

function build_one {
	mkdir ${CPU}
	cd ${CPU}

	echo "Configuring..."
	cmake -DANDROID_NATIVE_API_LEVEL=${API} -DANDROID_ABI=${CPU} -DCMAKE_BUILD_TYPE=Release -DANDROID_NDK=${NDK} -DCMAKE_TOOLCHAIN_FILE=$NDK/build/cmake/android.toolchain.cmake -DANDROID_NATIVE_API_LEVEL=16 -GNinja -DCMAKE_MAKE_PROGRAM=${NINJA_PATH} ../..
	
	echo "Building..."
	cmake --build .
	
	cd ..
}

function checkPreRequisites {

    if ! [ -d "boringssl" ] || ! [ "$(ls -A boringssl)" ]; then
        echo -e "\033[31mFailed! Submodule 'boringssl' not found!\033[0m"
        echo -e "\033[31mTry to run: 'git submodule init && git submodule update'\033[0m"
        exit
    fi

    if [ -z "$NDK" -a "$NDK" == "" ]; then
        echo -e "\033[31mFailed! NDK is empty. Run 'export NDK=[PATH_TO_NDK]'\033[0m"
        exit
    fi
    
    if [ -z "$NINJA_PATH" -a "$NINJA_PATH" == "" ]; then
        echo -e "\033[31mFailed! NINJA_PATH is empty. Run 'export NINJA_PATH=[PATH_TO_NINJA]'\033[0m"
        exit
    fi
}

ANDROID_NDK=$NDK
checkPreRequisites

cd boringssl

mkdir build
cd build

API=16

CPU=armeabi-v7a
build_one

CPU=x86
build_one

API=21
CPU=arm64-v8a
build_one

CPU=x86_64
build_one
